Main

jinusong opened the pull request:
추가 기능 구현은 아직인 페스타 메인입니다!

jinusong committed:
[Add] main xml

jinusong committed:
[Add] event list adapter

jinusong committed:
[Update] event domain impl

jinusong committed:
[Add] event data mapper

jinusong committed:
[Update] event presentation logic

jinusong committed:
[Update] internet permisson

jinusong committed:
[Delete] Api Client

jinusong committed:
[Add] event datasource

jinusong committed:
[Update] data entity

jinusong committed:
[Add] data mapper

jinusong committed:
[Update] event repository

jinusong committed:
[Update] evnet entity and impl event service

jinusong committed:
[Update] layout list gravity

jinusong committed:
[Update] di and presentation logic

kangkw commented on app/src/main/res/layout/item_event_list.xml:
```diff
@@ -0,0 +1,72 @@
+
+        <View
+                android:layout_margin="10dp"
+                android:background="#F4F4F4"
```
color 리소스로 빼주세요

kyeongwan commented on app/src/main/res/layout/item_event_list.xml:
```diff
@@ -0,0 +1,72 @@
+
+        <View
+                android:layout_margin="10dp"
+                android:background="#F4F4F4"
```
color 리소스로 빼주세요

jinusong replied:
알겠습니다

jinusong committed:
[Update] Add gray color to resource

moved gray color from the main layout to resource

kyeongwan commented:
ㅋㅋㅋㅋ 시간날때 천천히 해줄게요

jinusong commented:
> ㅋㅋㅋㅋ 시간날때 천천히 해줄게요

넵, 코드리뷰 해주시는 것만으로도 정말 감사해요 ㅠㅠ

jinusong committed:
[Add] Add Room and Update Event data layer

event datasource and repository's data logic has been changed by applying room database.

jinusong committed:
[Update] Apply offline mode

Offline mode applied.
If the network connection fails,
the domain and presentation of
the event are changed by the logic that fetches the data from the database.

jinusong committed:
[Fix] EventList loading bug

There was a bug where usecase's observable became dispose
and usecase could not be used again.

I solved this bug by clearing the composite.

jinusong committed:
[Update]  Change type of EventList

Change the type of EventList from ArrayList to List.

The reason for this is that you have no code to use while using ArrayList.

jinusong committed:
[Update] Apply eventList DTO

I removed the pageData
which is an unnecessary entity in the data layer and added EventListDTO

jinusong committed:
[Fix] Event Key

Set the Event PrimaryKey

kyeongwan commented on app/src/main/java/com/jinwoo/festa/presentation/base/BasePresenter.kt:
```diff
@@ -4,23 +4,18 @@ import io.reactivex.disposables.CompositeDisposable
 import io.reactivex.disposables.Disposable
```
not use

jinusong replied:
사용하지 않는 import 문을 신경쓰지 않았네요... 수정하겠습니다

kyeongwan commented:
baseUrl 은 https://festa.io/api/ 까지만 사용하세요.
보통 버전은 도메인별로 달라질 수 있습니다.

kyeongwan commented on app/src/main/java/com/jinwoo/festa/presentation/main/MainPresenter.kt:
```diff
@@ -1,12 +1,62 @@
+        }
+    }
 
+    override fun destroyView() {
```
개인적으로는 override 문을 먼저 두고 밑에 local function 을 두는걸 선호하는데 이건 취향차이

jinusong replied:
알겠습니다

kyeongwan commented on app/src/main/java/com/jinwoo/festa/presentation/main/MainPresenter.kt:
```diff
@@ -1,12 +1,62 @@
+    }
+
+    private fun getEventList() {
+        if (view.checkNetwork()) getRemoteEventList()
```
네트워크 체크부분은 서비스단으로 빼서 거기서 에러로 흘려주면 어떨까요

jinusong replied:
종합적으로 보자면 네트워크 에러 체크는 서비스에서 한 후 서비스에서 레포지토리를 통해서 캐시를 저장하고 로드 하는 방향으로 가는게 불필요한 UseCase나 View에서 Boolean 값을 리턴해주는 걸 방지할 수 있다는 것이네요.

리뷰를 받고 생각해보니까 제가 봐도 서비스에서 처리하는게 맞는 것 같네요. 이런 구조를 처음 구현하다보니까 시행착오가 많은데 바로 잡아주셔서 감사합니다

jinusong replied:
> 네트워크 체크부분은 서비스단으로 빼서 거기서 에러로 흘려주면 어떨까요

종합적으로 보자면 네트워크 에러 체크는 서비스에서 한 후 서비스에서 레포지토리를 통해서 캐시를 저장하고 로드 하는 방향으로 가는게 불필요한 UseCase나 View에서 Boolean 값을 리턴해주는 걸 방지할 수 있다는 것이네요.

리뷰를 받고 생각해보니까 제가 봐도 서비스에서 처리하는게 맞는 것 같네요. 이런 구조를 처음 구현하다보니까 시행착오가 많은데 바로 잡아주셔서 감사합니다

jinusong replied:
commit: 625ad216138f3fdc00d027e0adf0dc221f3af464

수정되었습니다.
말씀하신대로 에러 핸들링은 서비스 단에서 처리하였고 캐시의 저장 및 가져오기의 흐름도 서비스에서 조절할 수 있게 하였습니다. 
또한, EventEntity가 remote의 대한 결과인지 local에 대한 결과인지 구분하기 위해 Entity에 state라는 Boolean값을 넣어주어 구분하도록 하였습니다.

kyeongwan commented on app/src/main/java/com/jinwoo/festa/presentation/main/MainPresenter.kt:
```diff
@@ -1,12 +1,62 @@
+                override fun onComplete() {
+                    view.createToast("조회 성공")
+                    CoroutineScope(Dispatchers.IO).launch {
+                        getEventListUseCase.saveLocalEventList(
```
캐시는 서비스에서 레포지토리에서 바로 저장하세요

jinusong replied:
이 것도 서비스에서 처리하면 굳이 프레젠터에서 코루틴을 사용할 필요가 없겠네요

kyeongwan commented on app/src/main/java/com/jinwoo/festa/presentation/main/MainPresenter.kt:
```diff
@@ -1,12 +1,62 @@
+                        getEventListUseCase.saveLocalEventList(
+                            ArrayList(eventList.map { eventModelMapper.mapModelToEntity(it) }))
+                    }
+                    view.setEventListAdapter(eventList)
```
set 보다는 view.show 정도의 네이밍이 더 깔끔할 것 같아요

kyeongwan replied:
그리고 아직은 안했지만 나중에 View 에서 들어온 이벤트로 Presenter 에 보낼때는 Event 가 발생했으니 on 으로 시작하는게 깔끔합니다

jinusong replied:
> set 보다는 view.show 정도의 네이밍이 더 깔끔할 것 같아요

데이터는 set되지만 리스트는 새로 show되니까 show가 맞을 수도 있을 것 같긴하네요.



> 그리고 아직은 안했지만 나중에 View 에서 들어온 이벤트로 Presenter 에 보낼때는 Event 가 발생했으니 on 으로 시작하는게 깔끔합니다

참고하겠습니다.

kyeongwan commented on app/src/main/java/com/jinwoo/festa/presentation/ui/MainActivity.kt:
```diff
@@ -1,13 +1,40 @@
+
+    override fun setEventListAdapter(models: ArrayList<EventModel>) {
+        main_event_list.adapter = EventListAdapter(models)
+        main_event_list.layoutManager = GridLayoutManager(this, 2)
```
init 문은 init 에서 미리 해주세요

jinusong replied:
init 문에서 충분히 미리 처리할 수 있는 코드였네요

kyeongwan commented on app/src/main/java/com/jinwoo/festa/presentation/ui/MainActivity.kt:
```diff
@@ -1,13 +1,40 @@
+        presenter.createView(this)
+    }
+
+    override fun checkNetwork(): Boolean {
```
프리젠터로 리턴하지 마세요

jinusong replied:
넵

kyeongwan commented on app/src/main/java/com/jinwoo/festa/presentation/ui/MainActivity.kt:
```diff
@@ -1,13 +1,40 @@
 
-class MainActivity : AppCompatActivity() {
+
+class MainActivity : BaseActivity(), MainContract.View {
```
ActivityScope?

jinusong replied:
그냥 Activity라서 Activity에 속하지 않으니까 ActivityScope는 안붙여도 되지 않나요?

kyeongwan commented on app/src/main/java/com/jinwoo/festa/presentation/ui/MainActivity.kt:
```diff
@@ -1,13 +1,40 @@
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
+        presenter.createView(this)
```
뷰도 대거로 던져버리고 생성자에서 받는게 어때요??
생명주기는 Base 에서 받고요

jinusong replied:
알겠습니다

kyeongwan commented on app/src/main/java/com/jinwoo/festa/presentation/main/MainPresenter.kt:
```diff
@@ -1,12 +1,62 @@
 package com.jinwoo.festa.presentation.main
 
+import android.util.Log
```
로그 지우기

jinusong replied:
앞으로 좀 더 신경쓰겠습니다....

jinusong replied:
commit: 366e0e96783ce60ec2e67c7a15a5a4c5bbfeb6c5

수정되었습니다!

kyeongwan commented on app/src/main/java/com/jinwoo/festa/presentation/mapper/EventModelMapper.kt:
```diff
@@ -0,0 +1,28 @@
+        hostName = from.hostName
+    )
+
+    fun mapModelToEntity(from: EventModel): EventEntity = EventEntity(
```
서비스에서 캐시 다 저장하고 나오면 다시 뷰모델이 엔터티로 매핑해서 돌아가는 일이 필요 없겠죠?

jinusong replied:
> 서비스에서 캐시 다 저장하고 나오면 다시 뷰모델이 엔터티로 매핑해서 돌아가는 일이 필요 없겠죠?

말씀하신 의도를 이해했고 수정하겠습니다

kyeongwan commented on app/src/main/java/com/jinwoo/festa/domain/usecase/GetEventListUseCase.kt:
```diff
@@ -0,0 +1,19 @@
+    fun getLocalEventList(): List<EventEntity>
+            = service.getLocalEventList()
+
+    fun saveLocalEventList(eventList: ArrayList<EventEntity>)
```
Usecase 하나당 하나의 통로만 열어주세요. Save 쓰고 싶으면
SaveEventListUseCase 클래스 하나 더 생성

jinusong reviewed (commented):
> baseUrl 은 https://festa.io/api/ 까지만 사용하세요.
> 보통 버전은 도메인별로 달라질 수 있습니다.

핑계를 대자면.... 
도메인 버전이 달라질 여지가 있는 baseurl은 처음이라서 신경을 못썼네요.
좀 더 신경쓰겠습니다.

jinusong committed:
[Delete] Not used import

Deleted import that I do not use.

jinusong committed:
[Update] Cache Data flow

Reduced the unnecessary code by treating
the data flow of the cache processed by Presenter as error handling in EventService.

jinusong committed:
[Update] Rename function name and fix UI error

Renamed function setEventListAdapter to showEventListAdapter.

Fixed itemEventList UI TextView Bug

jinusong committed:
[Update] code naming and structure feedback

Naming and layer roles update based on feedback received during mentoring.

kangkw commented:
아이고
